PR: `Phoenix` `LiveView` Drag-and-Drop Example

SimonLab opened the pull request:
Create Phoenix application
see: #4

SimonLab commented:
I've added Tailwind to the Phoenix application with https://github.com/dwyl/learn-alpine.js/pull/5/commits/c4f97433b89751d808d77c85d02975b0e291c033

However the hot reload when changing a css class doesn't seem to work. I have to stop and restart the server to be able to see the Tailwind changes. @nelsonic when you have some time, do you mind to try this branch to see if changing a css Tailwind class automatically apply the new style on your browser. I'm not sure if it's a configuration issue on my machine or on the Phx application itself.

nelsonic commented:
@SimonLab remember to add the "watcher". Find the `watchers` section in `config/dev.exs` and add:

```elixir
tailwind: {Tailwind, :install_and_run, [:default, ~w(--watch)]}
```

See: https://github.com/dwyl/learn-tailwind/blob/61886d3a9821873c02c2bae7cc3a42692b90ef58/config/dev.exs#L17-L21

SimonLab commented:
Yes I've added the watcher here:
https://github.com/dwyl/learn-alpine.js/blob/c4f97433b89751d808d77c85d02975b0e291c033/config/dev.exs#L27-L31

That's why I'm not sure what the css is not compiled automatically when I changed a class

nelsonic commented:
Hmm ... that's strange. Let me try and run it on my `localhost`. 👨‍💻 
If you want to get on Zoom to remote-pair/debug this LMK (via Signal). 👌

nelsonic commented:
@SimonLab I did:

```sh
git checkout 'phoenix-example-#4'
mix deps.get
mix assets.deploy
mix s
```
Works for me:


https://user-images.githubusercontent.com/194400/197215548-43edffda-f74d-4e1c-8088-dff5dd8352f3.mov

SimonLab commented:
Thanks for checking, I didn't run `mix assets.deploy` as I thought it was used only on production.
I'll try again now

SimonLab commented:
Weirdly hotreload works with Tailwind v3.18, however v3.2.0 or v3.2.1 doesn't change the UI on reload!
@nelsonic did you run `mix tailwind.install` to make sure you tested with v3.2.0?

I don't want to spend too much time on this specific issue, if it's working with 3.1.8 I'll continue with this and see later on if I can find what is the problem with v3.2

nelsonic commented:
@SimonLab yeah, it worked with `Tailwind v3.2.0` but totally fine to downgrade if it allows you to make progress. 👌

nelsonic commented:
@SimonLab looking great so far. Thanks for the demo during standup. 👌

SimonLab commented:
It took me a bit of time to implement a solution using javascript: https://github.com/dwyl/learn-alpine.js/pull/5/commits/eeba5be58495b990dde0b15a55f46c47823c3010

My goal is now to see how to implement the vanilla js logic using Alpine.js.
However I've noticed a few things I need to fix:
- The new item created are not displayed on connected clients, I need to make sure the liveview event is dispatch to all browsers
- Adding a new item break the js logic. I think it's because the dom as changed with liveview and js is not able to automatically see the new changes. I'll need to see how to work with LiveView and JS.

LuchoTurtle commented on drag-and-drop.md:
```diff
@@ -0,0 +1,780 @@
+We can use the [mix phx.gen.live](https://hexdocs.pm/phoenix/Mix.Tasks.Phx.Gen.Live.html)
+command:
+
+run `mix phx.gen.live Tasks Item items text:string index:integer`
```
I'm following this doc. When I was installing the Petal Components package, I noticed one of [their setup steps](https://petal.build/components) was 

```
You may have a modal/1 function defined already, which will cause a clash with PetalComponents.Modal.modal/1. Check in live_helpers.ex and remove it.
```

This only makes sense _after_ running this command. This might lead to confusion and a improper Petal set up for people that never did it 😥

LuchoTurtle replied:
In addition to this, I've reached this stage and, even after following Petal's guide and running this command, I can't seem to get a fresh project running. I get a 

```
** (CompileError) lib/app_web/views/page_view.ex: undefined function container/1 (expected AppWeb.PageView to define such a function or for it to be imported, but none are available)
```

Perhaps I'm missing something but I'm struggling to get it running 😞 . Perhaps it's me but I feel like something is missing here 👀

SimonLab commented:
@LuchoTurtle I've updated the documentation (you won't need to install Petal) and this should be easier to setup and follow but let me know if you still spot some issues, thanks

nelsonic commented:
```sh
git checkout 'phoenix-example-#4'
git pull
mix deps.get
mix ecto.reset
mix s
```

![drag-and-drop-demo](https://user-images.githubusercontent.com/194400/199440271-dee75a14-f699-4e36-8ab6-d19a504b295e.gif)

Still need to finish looking at the code in more detail. But so far so good. 👌

SimonLab commented:
I forgot to update the code to the latest version described in the doc. I'm going to do this soon! Thank for testing the PR!

LuchoTurtle commented:
Finished going through the guide and got it working splendidly! Thank you @SimonLab !

I'll leave the code review to the pros 😅

nelsonic commented on lib/app_web/live/item_live/index.html.heex:
```diff
@@ -0,0 +1,45 @@
+<h1>Listing Items</h1>
+
+<%= if @live_action in [:new, :edit] do %>
+  <.modal return_to={Routes.item_index_path(@socket, :index)}>
```
@SimonLab do we _need_ to use a `<.modal>` for creating `new` `items`? 🤷‍♂️

SimonLab replied:
The modal is created automatically by Phoenix when running the `phx.gen.live` command at https://github.com/dwyl/learn-alpine.js/blob/phoenix-example-%234/drag-and-drop.md#create-items

I thought about updating the way Phoenix creates the items however I didn't want to make the tutorial longer and preferred to focus on the drag and drop feature part itself.

nelsonic replied:
@SimonLab thanks for clarifying that this was generated by `phx.gen.live`. 👌 
It's _definitely_ concerning [to me] that someone in the Phoenix team 🤷‍♂️ 
thought it was "OK" to include **modals** in the default UX for creating new content. 🤦‍♂️ 
Modals are not even remotely "OK"; they're a UX anti-pattern 😢 
and a major red flag of lazy design. 🔴 

For the purposes of this example/spike, 🧑‍💻 
the objective is to demo drag-and-drop 🐉 
and you have achieved that. ✅ 

In terms of our scarce/valuable time we don't _need_ to spend any more on this. 💸 
I've made a note to write about it: https://github.com/dwyl/product-ux-research/issues/38 📝 
We will _completely_ avoid them in our **`App`**. 🙅
